### PR TITLE
Add PGP key retrieval endpoint

### LINKED ISSUE #2887: Allow Fleet Server to serve the PGP key when stored on disk
If Elastic Agent are unable to connect with Elastic public URL to retrieve the PGP key, they will fallback to a Fleet Server URL where the public key can be hosted too.

In Fleet Server, we must define a fixed route for hosting the GPG key `downloads/signing/key.pub`  to mirror the public URL and have it read the GPG key in from its configuration using the same secure mechanisms we use to read the TLS private key. 
Mandate HTTPS on this endpoint and let Elastic Agent call it whenever needed.

Related Elastic Agent issue - https://github.com/elastic/elastic-agent/issues/3264

### PR DESCRIPTION
**Pull request by michel-laterman**
## What is the problem this PR solves?

Fleet-server must be able to have a request path where a valid PGP key will be served to requesting agents.
This endpoint must only be valid when the server runs HTTPS.
An operator must be able to change the key that is served.

## How does this PR solve the problem?

Add a new endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key` that will return a PGP key.
The endpoint needs a valid API key and is rate limited.
Currently the `major`, `minor`, and `patch` values are ignored.

The endpoint will server a PGP key from it's cache if one exists, otherwise it will serve the contents of a PGP key found on disk at `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`. Note that the directory can be changed through configuration.
The key value is added to cache if it exists.

If this file does not exist the fleet-server will retrieve the key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` and serve it's contents (writing them to cache and disk). The upstream URL can also be changed through configuration.

## Design Checklist

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- ~~I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.~~
- [x] I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

## Related issues

- Closes #2887

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -260,6 +260,17 @@ the test BenchmarkXXXX and not unit tests)
 - `BENCHMARK_COUNT`: you can define the number of iterations go test will run. Having larger number helps
 remove run-to-run variations (Default: 8)
 
+#### E2E Tests
```
Comment (nchaulet) [inline README.md:263]: 🎉 nice to add some doc on how to run tests

**Comment by nchaulet** [comment]
I am wondering if we should read the key on start as we do for the the other filer we read from the config like (`SecretTokenPath` , ...) it will be less confusing for the user to change the key it will have to restart fleet-server and not rely on a cache expiring no?

File: internal/pkg/api/handlePGPRequest.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		return nil, err
+	}
+
+	// file does not exist
+	p, err = pt.getPGPFromUpstream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	pt.cache.SetPGPKey(key, p)
+	pt.writeKeyToDir(ctx, zlog, key, p)
```
Comment (nchaulet) [inline internal/pkg/api/handlePGPRequest.go:107]: If we write the upstream key to disk, this mean if that key is updated user will have to manually update it on their fleet-server?

Comment (michel-laterman) [reply internal/pkg/api/handlePGPRequest.go:107]: Yes, if the upstream key is updated the user will need to manually update the key on disk (or remove it)

File: internal/pkg/api/handlePGPRequest.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	if !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	// file does not exist
+	p, err = pt.getPGPFromUpstream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	pt.cache.SetPGPKey(key, p)
```
Comment (michel-laterman) [inline internal/pkg/api/handlePGPRequest.go:106]: @nchaulet If we wanted to make changing the key source easier we could use path+key when interacting with the cache.

If an operator needs to do an emergency release in this case just changing `pgp.dir` to a new path would immediately take effect (and either retrieve the key from disk or from a new upstream url if that's changed).
What do you think?

Comment (nchaulet) [reply internal/pkg/api/handlePGPRequest.go:106]: Yes I think it could be a better solution

**Review by nchaulet — APPROVED** [review]
Thanks for addressing the cache changes 🚀

**Pull request merged**